tests/bigquery/udf_tests.py

```diff
@@ -21,44 +20,81 @@
 
 class TestCases(unittest.TestCase):
 
-  def test_sql_building(self):
-    context = self._create_context()
-    table = google.datalab.bigquery.Table('test:requestlogs.today', context=context)
+  def test_building_udf(self):
+    code = 'console.log("test");'
+    return_type = 'integer'
+    params = {'test_param': 'integer'}
+    language = 'js'
+    imports = ['gcs://mylib']
+    udf = google.datalab.bigquery.UDF('test_udf', code, return_type, params, language, imports)
 
-    udf = self._create_udf()
-    query = google.datalab.bigquery.Query('SELECT * FROM foo($t)', t=table, udfs=[udf],
-                                          context=context)
+    expected_sql = '''\
+CREATE TEMPORARY FUNCTION test_udf (test_param integer)
+RETURNS integer
+LANGUAGE js
+AS """
+console.log("test");
+"""
+OPTIONS (
+library="gcs://mylib"
+);\
+'''
+    self.assertEqual(udf.name, 'test_udf')
+    self.assertEqual(udf.code, code)
+    self.assertEqual(udf.imports, imports)
 
-    expected_js = '\nfoo=function(r,emit) { emit({output1: r.field2, output2: r.field1 }); };\n' +\
-                  'bigquery.defineFunction(\'foo\', ["field1", "field2"], ' +\
-                  '[{"name": "output1", "type": "integer"}, ' +\
-                  '{"name": "output2", "type": "string"}], foo);'
-    self.assertEqual(query.sql, 'SELECT * FROM '
-                                '(SELECT output1, output2 FROM foo(`test:requestlogs.today`))')
-    self.assertEqual(udf._code, expected_js)
+    self.assertEqual(udf._language, language)
+    self.assertEqual(udf._repr_sql_(), udf._expanded_sql())
+    self.assertEqual(udf.__repr__(), 'BigQuery UDF - code:\n%s' % udf._code)
+    self.assertEqual(udf._expanded_sql(), expected_sql)
 
-  def test_udf_expansion(self):
-    sql = 'SELECT * FROM udf(source)'
-    udf = google.datalab.bigquery.UDF('inputs', [('foo', 'string'), ('bar', 'integer')], 'udf',
-                                      'code')
-    context = TestCases._create_context()
-    query = google.datalab.bigquery.Query(sql, udf=udf, context=context)
-    self.assertEquals('SELECT * FROM (SELECT foo, bar FROM udf(source))', query.sql)
+    # bad return type
+    return_type = ['integer']
+    with self.assertRaises(TypeError):
+      google.datalab.bigquery.UDF('test_udf', code, return_type, params, language, imports)
+    return_type = 'integer'
```

Comment:
I don't think this is a good pattern, where you set some data to a bad value, run a test, then set the data back to a good value for the next test. In particular with this series of tests, because you are asserting there should be a TypeError but with no further details, if you happen to make a mistake setting the value back to a good value, you may not be testing what you expect, as the code may be failing due to your incorrectly set "good" values rather than the expected bad value. This is the reasoning behind the general recommentation to run only a single test in each testing function, so that all state will be reset and you don't have to worry about state from one test contaminating the next test.

One possible approach here would be to write a helper function that creates an object that has all of the fields filled out with good data:
  d = initWithGoodData()  // I'm sure you can come up with a better name
  d.return_type = ['integer']   // bad data for this field
  with self.assertRaisesRegexp(TypeError, 'return_type'):
    google.datalab.bigquery.UDF('test_udf', d.code, d.return_type, d.params, d.language, d.imports);
Then you can put each of these tests into a separate test function.